2022-09-28-introduce-q-d.md

```diff
@@ -0,0 +1,241 @@
+
+> • Conflict-Free Liveness (Head):
+>
+> In presence of a network adversary, a conflict-free execution satisfies the following condition:
+> For any transaction tx input via (new,tx), tx ∈ T i∈[n] Ci eventually holds.
+
+This property and similar ones are encoded as a _Dynamic Logic_ expressions, and a suitable Model of a Hydra network is defined as an instance of `StateModel` from which test sequences representing User actions are generated.
+
+Hydra is a distributed system where nodes are interconnected through a network layer, and each node needs to be connected to a cardano-node in order to preserve the security of the protocol. While testing an actual "cluster" of hydra and cardano nodes is definitely possible, and certainly desirable at some point in order to strengthen confidence in the whole system, it would also be very slow: Spinning up processes, establishing network connections, producing blocks on a chain, all take seconds if not minutes which makes any signficant exploration of the model state space practically infeasible.
+
+Generated test traces are run within the [IOSim](https://github.com/input-output-hk/hydra-poc/blob/master/hydra-node/test/Hydra/ModelSpec.hs#L220) monad which allows testing 100s of traces within seconds.
+
+Of course, this means we won't be using real TCP/IP networking stack nor connection to a real Cardano node and chain to create a Hydra network, but this is actually not a liability but an asset. By [mocking](https://abailly.github.io/posts/mocking-good-idea.html) the interfaces Hydra nodes use to communicate with other nodes and Cardano network, we are able to control the behaviour of the communication layer and _inject faults_ representing some _Adversary_: Reordering or dropping messages, tampering the data, delaying requests...
+
+## Principles
+
+We'll use the latter example to illustrate quickcheck-dynamic's principles and give the reader an intuition on the four steps that need to be defined in order to use it: Defining a test _Model_, stating how the model relates to the _Implementation_, expressing _Properties_ and, last but not least, checking properties.
+
+### Defining a Model
+
+In quickcheck-dynamic, a _Model_ is some type, a representation of the expected state of the system-under-test, for which there exists an instance of the [StateModel class](https://github.com/input-output-hk/quickcheck-dynamic/blob/abailly-iohk/blog-post/quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs#L56) which sets the building blocks needed to generate and validate test sequences.
+
+In the case of Hydra, the Model is a `WorldState` data type that control the Head parties and maintains a `GlobalState` which reflects the expected Head state:
+
+```haskell
+data WorldState = WorldState
```

Comment:
Maybe we should call this `IdealWorld` to highlight that it models an ideal world situation? And an alternative (future) model will be `NetworkAdversaryWorld`?

Reply:
That's a good idea.